# 5.1.10 Введение в обработку исключений. Блоки try / except

Подвиг 10. Объявите в программе класс `FloatValidator`, объекты которого создаются командой:

```python
fv = FloatValidator(min_value, max_value)
```

где `min_value, max_value` - минимальное и максимальное допустимое значение (диапазон `[min_value; max_value]`).

Объекты этого класса предполагается использовать следующим образом:

```python
fv(value)
```

где `value` - проверяемое значение. Если `value` не вещественное число или не принадлежит
диапазону `[min_value; max_value]`, то генерируется исключение командой:

```python
raise ValueError('значение не прошло валидацию')
```

По аналогии, объявите класс `IntegerValidator`, объекты которого создаются командой:

```python
iv = IntegerValidator(min_value, max_value)
```

и используются командой:

```python
iv(value)
```

Здесь также генерируется исключение:

```python
raise ValueError('значение не прошло валидацию')
```

если `value` не целое число или не принадлежит диапазону `[min_value; max_value]`.

После этого объявите функцию с сигнатурой:

```python
def is_valid(lst, validators): ...
```

где `lst` - список из данных; `validators` - список из объектов-валидаторов (объектов классов `FloatValidator`
и `IntegerValidator`).

Эта функция должна отбирать из списка все значения, которые прошли хотя бы по одному валидатору. И возвращать новый
список с элементами, прошедшими проверку.

Пример использования классов и функции (эти строчки в программе не писать):

```python
fv = FloatValidator(0, 10.5)
iv = IntegerValidator(-10, 20)
lst_out = is_valid([1, 4.5, -10.5, 100, True, 'abc', (1, 2)], validators=[fv, iv])  # [1, 4.5]
```

P.S. В программе нужно только объявить классы и функцию. На экран ничего выводить не нужно.

# Solution

```python
class FloatValidator:
    _type = float

    def __init__(self, min_value, max_value):
        self.min_value, self.max_value = min_value, max_value

    def __call__(self, value):
        if type(value) != self._type or not self.min_value <= value <= self.max_value:
            raise ValueError('значение не прошло валидацию')
        return value


class IntegerValidator(FloatValidator):
    _type = int


def is_valid(lst, validators):
    res = []
    for item in lst:
        for val in validators:
            try:
                res.append(val(item))
            except:
                continue
    return res
```